elements/lisk-p2p/src/utils.ts

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { hash } from '@liskhq/lisk-cryptography';
+import { isIPv4 } from 'net';
+
+const SECRET_BUFFER_LENGTH = 4;
+const NETWORK_BUFFER_LENGTH = 1;
+const PREFIX_BUFFER_LENGTH = 1;
+const BYTES_4 = 4;
+const BYTES_16 = 16;
+const BYTES_64 = 64;
+const BYTES_128 = 128;
+const NEW_PEERS = 'new';
+const TRIED_PEERS = 'tried';
+
+export enum NETWORK {
+	NET_IPV4 = 0,
+	NET_PRIVATE,
+	NET_LOCAL,
+	NET_OTHER,
+}
+
+/* tslint:disable no-magic-numbers */
+export const getGroup = (
+	address: string,
+	groupNumber: number,
+): number | undefined => {
+	if (groupNumber > 3) {
+		return undefined;
+	}
+
+	const group = parseInt(address.split('.')[groupNumber], 10);
+
+	if (!Number.isInteger(group)) {
+		return undefined;
+	}
+
+	return group;
+};
+
+interface AddressBytes {
+	readonly aBytes: Buffer;
+	readonly bBytes: Buffer;
+	readonly cBytes: Buffer;
+	readonly dBytes: Buffer;
+}
+
+export const getAddressBytes = (address: string): AddressBytes => {
+	const aBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	aBytes.writeUInt8(getGroup(address, 0) as number, 0);
+	const bBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	bBytes.writeUInt8(getGroup(address, 1) as number, 0);
+	const cBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	cBytes.writeUInt8(getGroup(address, 2) as number, 0);
+	const dBytes = Buffer.alloc(PREFIX_BUFFER_LENGTH);
+	dBytes.writeUInt8(getGroup(address, 3) as number, 0);
+
+	return {
+		aBytes,
+		bBytes,
+		cBytes,
+		dBytes,
+	};
+};
+
+export const isPrivate = (address: string) =>
+	getGroup(address, 0) === 10 ||
+	(getGroup(address, 0) === 172 &&
+		((getGroup(address, 1) as number) >= 16 ||
+			(getGroup(address, 1) as number) <= 31));
+
+export const isLocal = (address: string) =>
+	getGroup(address, 0) === 127 || getGroup(address, 0) === 0;
+/* tslint:enable no-magic-numbers */
+
+export const getNetwork = (address: string): NETWORK => {
+	if (isLocal(address)) {
+		return NETWORK.NET_LOCAL;
+	}
+
+	if (isPrivate(address)) {
+		return NETWORK.NET_PRIVATE;
+	}
+
+	if (isIPv4(address)) {
+		return NETWORK.NET_IPV4;
+	}
+
+	return NETWORK.NET_OTHER;
+};
+
+interface GroupBytes {
+	readonly secretBytes: Buffer;
+	readonly networkBytes: Buffer;
+	readonly targetABytes: Buffer;
+	readonly targetBBytes: Buffer;
+	readonly targetCBytes: Buffer;
+	readonly targetDBytes: Buffer;
+	readonly sourceABytes?: Buffer;
+	readonly sourceBBytes?: Buffer;
+}
+
+export const getGroupBytes = (options: {
```

Comment:
It's difficult to describe in words what this function does. Maybe it shouldn't exist. It's a small function and half of it is potentially cut off by an if condition.

Reply:
There was some overlap in logic between `getNetgroup` and `getBucket` which this function covered. I agree it's hard to verbalize the purpose of it. In this case maybe it's better to have a small overlap to make it easier to understand.